File: rustls/src/verify.rs
/*
 * Copyright (C) 2026, Stephan Mueller <smueller@chronox.de>
 *
 * License: see LICENSE file in root directory
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE, ALL OF
 * WHICH ARE HEREBY DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
 * OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF NOT ADVISED OF THE POSSIBILITY OF SUCH
 * DAMAGE.
 */

use core::fmt;
use leancrypto_sys::{lcr_dilithium::lcr_dilithium, lcr_ed25519::lcr_ed25519};
use rustls::pki_types::alg_id;

use rustls::{
    SignatureScheme,
    crypto::WebPkiSupportedAlgorithms,
    pki_types::{
        AlgorithmIdentifier, InvalidSignature, SignatureVerificationAlgorithm,
    },
};

/// A [WebPkiSupportedAlgorithms] value defining the supported signature algorithms.
pub static SUPPORTED_SIG_ALGS: WebPkiSupportedAlgorithms =
    WebPkiSupportedAlgorithms {
        all: &[ML_DSA_87, ML_DSA_65, ML_DSA_44, ED25519],
        mapping: &[
            (SignatureScheme::ML_DSA_87, &[ML_DSA_87]),
            (SignatureScheme::ML_DSA_65, &[ML_DSA_65]),
            (SignatureScheme::ML_DSA_44, &[ML_DSA_44]),
            #[cfg(feature = "nonpqc")]
            (SignatureScheme::ED25519, &[ED25519]),
        ],
    };

/// ML-DSA-87
pub(crate) static ML_DSA_87: &dyn SignatureVerificationAlgorithm =
    &LeancryptoAlgorithm {
        display_name: "ML_DSA_87",
        public_key_alg_id: alg_id::ML_DSA_87,
        signature_alg_id: alg_id::ML_DSA_87,
    };

/// ML-DSA-65
pub(crate) static ML_DSA_65: &dyn SignatureVerificationAlgorithm =
    &LeancryptoAlgorithm {
        display_name: "ML_DSA_65",
        public_key_alg_id: alg_id::ML_DSA_65,
        signature_alg_id: alg_id::ML_DSA_65,
    };

/// ML-DSA-44
pub(crate) static ML_DSA_44: &dyn SignatureVerificationAlgorithm =
    &LeancryptoAlgorithm {
        display_name: "ML_DSA_44",
        public_key_alg_id: alg_id::ML_DSA_44,
        signature_alg_id: alg_id::ML_DSA_44,
    };

/// ED25519
#[cfg(feature = "nonpqc")]
pub(crate) static ED25519: &dyn SignatureVerificationAlgorithm =
    &LeancryptoAlgorithm {
        display_name: "ED25519",
        public_key_alg_id: alg_id::ED25519,
        signature_alg_id: alg_id::ED25519,
    };

struct LeancryptoAlgorithm {
    display_name: &'static str,
    public_key_alg_id: AlgorithmIdentifier,
    signature_alg_id: AlgorithmIdentifier,
}

impl fmt::Debug for LeancryptoAlgorithm {
    fn fmt(
        &self,
        f: &mut fmt::Formatter<'_>,
    ) -> fmt::Result {
        write!(
            f,
            "rustls_leancrypto Signature Verification Algorithm: {}",
            self.display_name
        )
    }
}

impl SignatureVerificationAlgorithm for LeancryptoAlgorithm {
    fn public_key_alg_id(&self) -> AlgorithmIdentifier {
        self.public_key_alg_id
    }

    fn signature_alg_id(&self) -> AlgorithmIdentifier {
        self.signature_alg_id
    }

    fn verify_signature(
        &self,
        public_key: &[u8],
        message: &[u8],
        signature: &[u8],
    ) -> Result<(), InvalidSignature> {
        match self.public_key_alg_id {
            alg_id::ML_DSA_44 | alg_id::ML_DSA_65 | alg_id::ML_DSA_87 => {
                let mut dilithium = lcr_dilithium::new();
                dilithium
                    .pk_load(&public_key)
                    .map_err(|_| InvalidSignature)?;
                dilithium
                    .sig_load(&signature)
                    .map_err(|_| InvalidSignature)?;
                dilithium.verify(&message).map_err(|_| InvalidSignature)?;
            }
            #[cfg(feature = "nonpqc")]
            alg_id::ED25519 => {
                let mut ed25519 = lcr_ed25519::new();
                ed25519.enable().map_err(|_| InvalidSignature)?;
                ed25519.pk_load(&public_key).map_err(|_| InvalidSignature)?;
                ed25519.sig_load(&signature).map_err(|_| InvalidSignature)?;
                ed25519.verify(&message).map_err(|_| InvalidSignature)?;
            }
            _ => todo!(),
        };

        Ok(())
    }

    fn fips(&self) -> bool {
        crate::fips::enabled()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_leancrypto_algorithm_debug() {
        #[cfg(feature = "nonpqc")]
        assert_eq!(
            format!("{:?}", ED25519),
            "rustls_leancrypto Signature Verification Algorithm: ED25519"
        );
        assert_eq!(
            format!("{:?}", ML_DSA_87),
            "rustls_leancrypto Signature Verification Algorithm: ML_DSA_87"
        );
        assert_eq!(
            format!("{:?}", ML_DSA_65),
            "rustls_leancrypto Signature Verification Algorithm: ML_DSA_65"
        );
        assert_eq!(
            format!("{:?}", ML_DSA_44),
            "rustls_leancrypto Signature Verification Algorithm: ML_DSA_44"
        );
    }
}

File: sntrup/src/avx2/verify.c
/*
 * Copyright (C) 2026, Stephan Mueller <smueller@chronox.de>
 *
 * License: see LICENSE file in root directory
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE, ALL OF
 * WHICH ARE HEREBY DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
 * OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF NOT ADVISED OF THE POSSIBILITY OF SUCH
 * DAMAGE.
 */
/*
 * The SNTRUP code is derived in parts from the code base
 * https://libntruprime.cr.yp.to/ which uses the following license:
 *
 * Copyright (C) 2024 Free Software Foundation, Inc.; This is free software;
 * see the source for copying conditions. There is NO; warranty; not even for
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * SPDX-License-Identifier: LicenseRef-PD-hp OR CC0-1.0 OR 0BSD OR MIT-0 OR MIT
 */

#include "ext_headers_x86.h"
#include "sntrup_int16.h"
#include "params.h"

int sntrup_verify_clen_avx2(const uint8_t *x, const uint8_t *y)
{
	__m256i diff = _mm256_set1_epi8(0);
	int differentbits = 0;
	int i = sntrup_verify_BYTES - 32;

	LC_FPU_ENABLE;

	for (;;) {
		do {
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wcast-align"
			__m256i x0 = _mm256_loadu_si256((__m256i *)x);
			__m256i y0 = _mm256_loadu_si256((__m256i *)y);
#pragma GCC diagnostic pop
			diff |= x0 ^ y0;
			i -= 32;
			x += 32;
			y += 32;
		} while (i >= 0);
		if (i <= -32)
			break;
		x += i;
		y += i;
	}

	diff |= _mm256_srli_epi16(diff, 8);
	diff |= _mm256_srli_epi32(diff, 16);
	diff |= _mm256_srli_epi64(diff, 32);

	differentbits = _mm256_extract_epi8(diff, 0);
	differentbits |= _mm256_extract_epi8(diff, 8);
	differentbits |= _mm256_extract_epi8(diff, 16);
	differentbits |= _mm256_extract_epi8(diff, 24);

	LC_FPU_DISABLE;

	return sntrup_int16_nonzero_mask((int16_t)differentbits);
}
